### PR TITLE
feature: Add FairProjectConfig

### PR DESCRIPTION
**Pull request by ChristianTackeGSI**
Currently it has only one function: fair_get_git_version()

### CODE REVIEW

**Comment by ChristianTackeGSI** [comment]
Should we add a `--always` to the`git describe`?

**Review by dennisklein — COMMENTED** [review]
https://github.com/FairRootGroup/FairCMakeModules/runs/3138860144?check_suite_focus=true#step:5:26 Is this behavior intended?

**Comment by dennisklein** [comment]
> Should we add a `--always` to the`git describe`?

To be honest, I do not really understand what it does.

**Comment by ChristianTackeGSI** [comment]
> https://github.com/FairRootGroup/FairCMakeModules/runs/3138860144?check_suite_focus=true#step:5:26 Is this behavior intended?

I guess, the problem is, that the github CI does not present a full git repository to the build process.

So we could either say: 
→ Okay, that's the case, where we want "don't touch variables" behaviour.

If that's the idea, we might consider to add `ERROR_VARIABLE` to `execute_process()` to hide the error messages. (I am unsure, if this is a good idea.)

Or:

> > Should we add a `--always` to the`git describe`?
> 
> To be honest, I do not really understand what it does.

If `git describe` can't find a good tag, it will return a shortened commit sha1 with that flag.

I have added a commit to show the effect:

```console
-- FairCMakeModules Version 0.2.0 (6c08dde)
```

For the github CI, we get some strange commit sha1. (I guess, it's some merge commit or something like that).

I am also unsure, whether this is really helpful. Maybe it's just "better than nothing"?

**Comment by dennisklein** [comment]
> I am also unsure, whether this is really helpful. Maybe it's just "better than nothing"?

What would the code look like, if a project wanted to implement the behaviour "Get the git version, but fall back to something else, if a git version is not available"? I believe we should make sure that such a code snippet is not going to get overly complicated / un-readable because of the interface we choose for `fair_get_git_version()`.

**Comment by dennisklein** [comment]
> > I am also unsure, whether this is really helpful. Maybe it's just "better than nothing"?
> 
> What would the code look like, if a project wanted to implement the behaviour "Get the git version, but fall back to something else, if a git version is not available"? I believe we should make sure that such a code snippet is not going to get overly complicated / un-readable because of the interface we choose for `fair_get_git_version()`.

```cmake
fair_get_git_version()
if(NOT PROJECT_GIT_VERSION)
  # get it from somewhere else
endif()
```

If there is no such logic in the project itself, and a package manager cannot present a git repo, it should also be straightforward to *inject* the version by modifying the source tree, e.g. with a package-level patch:

```diff
- fair_get_git_version()
+ set(PROJECT_GIT_VERSION "v42")
```

Looks simple enough, so from my side I would approve this.

> Maybe it's just "better than nothing"?

I would conclude, yes.

**Comment by ChristianTackeGSI** [comment]
If we don't implement `--always`:
```cmake
set(PROJECT_GIT_VERSION "your fallback default here")
fair_get_git_version()

# Or
fair_get_git_version(OUTVAR_PREFIX yourprefix)
if(NOT DEFINED yourprefix_GIT_VERSIONyourprefix_GIT_VERSION)
   set(yourprefix_GIT_VERSION "your fallback")
   # Do something else
endif()
```

For the `--always` case the above code would only work for a non-existant / broekn `.git`. Otherwise it would always use the sha1.

In that case, one would do something like
```cmake
fair_get_git_version()
if(NOT DEFINED PROJECT_GIT_VERSION OR PROJECT_GIT_VERSION regex-match-operator "^[0-9a-z][0-9a-z][0-9a-z][0-9a-z]*$")
  # ...
endif()
```

Which obviously is ugly.

**Comment by dennisklein** [comment]
> If that's the idea, we might consider to add ERROR_VARIABLE to execute_process() to hide the error messages. (I am unsure, if this is a good idea.)

Actually, what about having a flag `FATAL_ERROR` (or `REQUIRED`) which is by default `OFF` - meaning it would silence the error message and otherwise turn the output of `git describe` into a fatal cmake error on non-zero exit codes (possibly with some additional comments).

File: src/modules/FairProjectConfig.cmake [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  if(NOT ARGS_OUTVAR_PREFIX)
+    set(ARGS_OUTVAR_PREFIX PROJECT)
+  endif()
+
+  if(NOT ARGS_TAGS_ONLY)
+    set(git_always "--always")
+  endif()
+
+  if(GIT_FOUND AND EXISTS ${CMAKE_SOURCE_DIR}/.git)
```
Comment (dennisklein) [inline src/modules/FairProjectConfig.cmake]:
```suggestion
  if(Git_FOUND AND EXISTS ${CMAKE_SOURCE_DIR}/.git)
```

The find module also sets `GIT_FOUND` but `Git_FOUND` is the more modern convention.

File: src/modules/FairProjectConfig.cmake [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    execute_process(COMMAND ${GIT_EXECUTABLE} describe --tags --dirty ${git_always} --match "v*"
```
Comment (dennisklein) [inline src/modules/FairProjectConfig.cmake:83]:
```suggestion
    # TODO use this form once this module requires CMake 3.14
    # execute_process(COMMAND $<TARGET_FILE:Git::Git> describe --tags --dirty ${git_always} --match "v*"
    execute_process(COMMAND ${GIT_EXECUTABLE} describe --tags --dirty ${git_always} --match "v*"
```

File: src/modules/FairProjectConfig.cmake (hunk 1 above)
Comment (dennisklein) [reply src/modules/FairProjectConfig.cmake]: the existance check for `.git` is fragile, consider using `git rev-parse --is-inside-work-tree` or similar (not sure how recent the `--is-inside-work-tree` flag is, also see e.g. https://remarkablemark.org/blog/2020/06/05/check-git-repository/ for simpler variant)

**Review by dennisklein — CHANGES_REQUESTED** [review]

File: src/modules/FairProjectConfig.cmake (hunk 1 above)
Comment (ChristianTackeGSI) [reply src/modules/FairProjectConfig.cmake]: What about dropping the existence check completely and letting `git describe` handle it?

```console
CMake Error at src/modules/FairProjectConfig.cmake:91 (message):
  git describe failed (return value 128):

  fatal: not a git repository (or any parent up to mount point /)

  Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

**Review by dennisklein — APPROVED** [review]

**Pull request merged**